src/GHCMod/Options.hs

```diff
@@ -0,0 +1,174 @@
+    (long "version" <> help "Print the version of the program.") <*>
+  argument r (value id <> metavar "")
+  where
+    r :: ReadM (a -> a)
+    r = do
+      v <- readerAsk
+      case v of
+        "help" -> readerAbort ShowHelpText
+        "version" -> readerAbort $ InfoMsg ghcModVersion
+        _ -> return id
+
+argAndCmdSpec :: Parser (Options, GhcModCommands)
+argAndCmdSpec = (,) <$> globalArgSpec <*> commandsSpec
+
+splitOn :: Eq a => a -> [a] -> ([a], [a])
+splitOn c = second (drop 1) . break (==c)
+
+getLogLevel :: Int -> GmLogLevel
+getLogLevel = toEnum . min 7
+
+logLevelParser :: Parser GmLogLevel
+logLevelParser =
+  getLogLevel <$>
+    (
+    silentSwitch <|> logLevelSwitch <|> logLevelOption
+    )
+  where
+    logLevelOption =
+      option int (
+        long "verbose" <>
+        metavar "LEVEL" <>
+        value 4 <>
+        showDefault <>
+        help "Set log level. (0-7)"
+      )
+    logLevelSwitch =
+      (4+) . length <$> many (flag' () (
+        short 'v' <>
+        help "Increase log level"
+      ))
+    silentSwitch = flag' 0 (
+        long "silent" <>
+        short 's' <>
+        help "Be silent, set log level to 0"
+      )
+
+outputOptsSpec :: Parser OutputOpts
+outputOptsSpec = OutputOpts <$>
+      logLevelParser <*>
+      flag PlainStyle LispStyle (
+        long "tolisp" <>
+        short 'l' <>
+        help "Format output as an S-Expression"
+      ) <*>
+      (LineSeparator <$> strOption (
+        long "boundary" <>
+        long "line-separator" <>
+        short 'b' <>
+        metavar "SEP" <>
+        value "\0" <>
+        showDefault <>
+        help "Output line separator"
+      )) <*>
+      optional (splitOn ',' <$> strOption (
+        long "line-prefix" <>
+        metavar "OUT,ERR" <>
+        help "Output prefixes"
+      ))
+
+programsArgSpec :: Parser Programs
+programsArgSpec = Programs <$>
+      strOption (
+        long "with-ghc" <>
+        value "ghc" <>
+        showDefault <>
+        help "GHC executable to use"
+      ) <*>
+      strOption (
+        long "with-ghc-pkg" <>
+        value "ghc-pkg" <>
+        showDefault <>
+        help "ghc-pkg executable to use (only needed when guessing from GHC path fails)"
+      ) <*>
+      strOption (
+        long "with-cabal" <>
+        value "cabal" <>
+        showDefault <>
+        help "cabal-install executable to use"
+      ) <*>
+      strOption (
+        long "with-stack" <>
+        value "stack" <>
+        showDefault <>
+        help "stack executable to use"
+      )
+
+globalArgSpec :: Parser Options
+globalArgSpec = Options <$>
+      outputOptsSpec <*> -- optOutput
+      programsArgSpec <*> -- optPrograms
+      many (strOption ( -- optGhcUserOptions
+        long "ghcOpt" <>
+        long "ghc-option" <>
+        short 'g' <>
+        metavar "OPT" <>
+        help "Option to be passed to GHC"
+      )) <*>
+      many fileMappingSpec -- optFileMappings   = []
+  where
+    {-
```

Comment:
Well, I didn't find a way to make that presentable. For commands, it works well enough, but for options it sadly doesn't work out that great.